text/0002-new-architecture.md

```diff
@@ -0,0 +1,243 @@
+item types this can turn more into a serverless function like internal API for adding item types that make it easier
+to land new items.
+
+### Relay Disk Buffer
+
+Relays currently buffer all data in memory which means that if they go down, we lose data.  Instead relay ideally buffers
+all its data additionally in a Kafka topic or on-disk storage that can make recovery from extended downtime possible.
+
+### Partial Grouping Awareness in Relay
+
+Today our grouping system really can only run in the tail end of the processing pipeline.  This makes it impossible for
+us to perform certain types of filtering or grouping decisions in relay.  There is a range of issues that could already
+be grouped quite accurately in Relay which would then permit us to do metrics extraction in relay or to perform efficent
+discarding.
+
+### Tail-based Sampling / Distributed Sampling Context Buffer
+
+Relay is currently unable to perform tail based sampling or consistent sampling if the dynamic sampling context changes.
+This could be remedied by extensive buffering or alternative approaches.  While a full tail based sampling approach is
+likely to be quite costly, there might be hybrid approaches possible.
+
+## Service Infrastructure
+
+These are changes to the Sentry architecture to better support the creation of independent services.
+
+### In-Monolith Service Compartmentalization
+
+The Sentry monolith currently houses a range of small services that are however clustered together.  For instance the
+processing pipeline really could be separated entirely from the rest of the System.  However the same is true for quite
+a lot of different parts of the product.  For instance certain parts of the user experience are sufficiently idependent
+in the UI code today already that they could turn into a separate service that just supplies UI views together with some
+APIs.  While it's unclear if the existing experiences are worth splitting up into new services, it's probably likely that
+we can at least structure their code so that we can reduce the total amount of pieces of code that need to be checked and
+tested against changes.
+
+In the ideal situation a change to the settings page of a project for instance does not need to run the UI tests for
+performance views etc. unless the settings are in fact related to that component.
+
+### Service Declarations
+
+The biggest limiting factor today in creating new services is the configuration of these services.  When an engineer adds
+a new service there is typically a followup PR against the internal ops repo to configure these services for the main
+sentry installation.  Then there is some sort of ad-hoc workaround to make the whole thing work in local development, how
+to make it work on a single tenant or self hosted installation.  Each new services undergoes a new phase of rediscovery how
+suboptimal the rules and policies around scaling these services are.
```

Comment:
Are we overloading 'services'? Do we need a different term for the processes that a product service needs in order to operate?

Reply:
Agreed we should probably be more explicit about what those things are.

Reply:
semantic nit maybe. Should we start talking about domain boundaries and bounded contexts (from domain driven design) instead of services ? 
Defining bounded contexts is a logical partition of the data model (which requires having a coherent data model to begin with). That then implies clear dependencies between features/components as your component cannot directly access the part of the data model that belongs to another bounded context. 
Having the clear dependencies then allows us to isolate what we call services today whether or not we split the code base or not.

Reply:
> Should we start talking about domain boundaries and bounded contexts (from domain driven design) instead of services ?
Defining bounded contexts is a logical partition of the data model (which requires having a coherent data model to begin with). 

Totally agree that the first step to separating 'services' within the monolith or outside of it should start with bounded contexts and what consistency guarantees we need within each context.

Reply:
For the service separation within the monolith I'm trying to work something sensible out as a near term roadmap.